src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -433,69 +434,64 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
         {
-            var di = new DirectoryInfo(fullPath);
-            if (!di.Exists)
+            if (!TryRemoveDirectory(fullPath, topLevel: true, throwWhenNotEmpty: !recursive))
             {
-                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath, isDirectory: true);
+                RemoveDirectoryRecursive(fullPath);
             }
-            RemoveDirectoryInternal(di, recursive, throwOnTopLevelDirectoryNotFound: true);
         }
 
-        private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recursive, bool throwOnTopLevelDirectoryNotFound)
+        private static void RemoveDirectoryRecursive(string fullPath)
         {
             Exception? firstException = null;
 
-            if ((directory.Attributes & FileAttributes.ReparsePoint) != 0)
+            try
             {
-                DeleteFile(directory.FullName);
-                return;
-            }
+                foreach ((string childPath, bool isDirectory) in
+                            new FileSystemEnumerable<(string, bool)>(
+                                fullPath,
+                                static (ref FileSystemEntry entry) =>
+                                {
+                                    // Don't report symlinks to directories as directories.
+                                    bool isRealDirectory = !entry.IsSymbolicLink && entry.IsDirectory;
 
-            if (recursive)
-            {
-                try
+                                    return (entry.ToFullPath(), isRealDirectory);
+                                },
+                                EnumerationOptions.Compatible))
                 {
-                    foreach (string item in Directory.EnumerateFileSystemEntries(directory.FullName))
+                    try
                     {
-                        if (!ShouldIgnoreDirectory(Path.GetFileName(item)))
+                        if (isDirectory)
                         {
-                            try
-                            {
-                                var childDirectory = new DirectoryInfo(item);
-                                if (childDirectory.Exists)
-                                {
-                                    RemoveDirectoryInternal(childDirectory, recursive, throwOnTopLevelDirectoryNotFound: false);
-                                }
-                                else
-                                {
-                                    DeleteFile(item);
-                                }
-                            }
-                            catch (Exception exc)
-                            {
-                                if (firstException != null)
-                                {
-                                    firstException = exc;
-                                }
-                            }
+                            RemoveDirectoryRecursive(childPath);
```

Comment:
Would it be possible, to perform a single recursive enumeration using a single `FileSystemEnumerable` instance and avoid the need of having to call the method recursively?

Reply:
We can't do that because we need to delete from the leaves to the top.